### PR TITLE
Don't fail when What4 sends us a function-based array in a result.

### LINKED ISSUE #2120: Gracefully print counterexamples involving SMT arrays defined as function mappings
Generally speaking, SMT solvers produce models of SMT arrays in one of two ways when producing counterexamples:

1. A constant array where all indices map to a particular value, but with elements at particular indices updated to other values.
2. An array defined as a total function from index values to element values.

SAW is capable of printing SMT arrays of type (1) (e.g., `[<default> := False]`), but it is not capable of printing SMT arrays of type (2). Nor is the purpose of this issue to propose that SAW be able to print type-(2) arrays. Rather, the issue is that SAW's behavior when printing type-(2) arrays is currently rather poor. At present, attempting to do so will cause SAW to fail outright with a full-blown error message. This can be seen in the following example:

```
// test.saw
import "Array.cry";

prove_print w4
  {{ \(a : Array Integer Bool) (n : Integer) ->
     arrayLookup a n /\ ~(arrayLookup a (n + 1)) }};
```

With a recent SAW nightly, this will fail thusly:

```
[16:34:35.255] Loading file "/home/ryanscott/Documents/Hacking/SAW/test.saw"
[16:34:35.338] Stack trace:
"prove_print" (/home/ryanscott/Documents/Hacking/SAW/test.saw:4:1-4:12)
"w4" (/home/ryanscott/Documents/Hacking/SAW/test.saw:4:13-4:15)
Unsupported FOV Array (values only available by lookup)
```

(The "`Unsupported FOV Array (values only available by lookup)`" part of the error message is a very dense way of saying that it encountered a type-(2) SMT array.)

Really, we ought to gracefully handle these situations by printing these arrays as something opaque. This is more or less than SAW 1.2 and earlier do:

```
[16:34:46.127] Loading file "/home/ryanscott/Documents/Hacking/SAW/test.saw"
[16:34:46.222] Stack trace:
"prove_print" (/home/ryanscott/Documents/Hacking/SAW/test.saw:4:1-4:12)
prove: 1 unsolved subgoal(s)
Invalid: [a = FOTArray FOTInt FOTBit, n = 0]
```

The output "`FOTArray FOTInt FOTBit`" is rather obtuse, so we should consider something a bit prettier to print here, e.g., `<array Integer Bit>`. While we can't print out the full contents of a type-(2) array, we should at least print it out opaquely in the counterexample without causing the rest of SAW to fall over.

### PR DESCRIPTION
**Pull request by sauclovian-g**
(see commit for description)

Fixes #2120.

### CODE REVIEW

File: saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs [hunk 1]
```diff
@@ -114,10 +114,31 @@ groundToFOV BaseRealRepr    _         = Left "Real is not FOV"
...
+    -- type, and use that channel to return the actual function
+    -- handle. (Including the function handle in a FirstOrderValue
+    -- value is problematic as things stand.) However, there's nothing
+    -- to be gained by doing this without a way to figure out what
+    -- key values to look at. And realistically, if we ever _do_
+    -- come up with a way to figure out what key values to look at,
+    -- it should be implemented in What4 so What4 never returns
+    -- ArrayMapping.
+    --
+    -- (See note in FiniteValue.hs where FirstOrderValue lives.)
```
Comment (RyanGlScott) [inline saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs:138]: To make this more obvious:

```suggestion
    -- (See Note [FOVArray] in saw-core:Verifier.SAW.FiniteValue.)
```

File: saw-core/src/Verifier/SAW/FiniteValue.hs [hunk 2]
```diff
@@ -78,13 +78,20 @@ data FirstOrderType
 -- The type argument of FOVArray is the key type; the value type is
 -- derivable from the default value, which is the second argument. The third
 -- argument is an assignment for all the entries that have non-default values.
+--
+-- FOVOpaqueArray is for arrays we get back from the solver only as a
+-- function call that one can use for lookups. We can't do anything
+-- with that (see notes below) so we just treat it as an opaque blob.
```
Comment (RyanGlScott) [inline saw-core/src/Verifier/SAW/FiniteValue.hs:84]:
```suggestion
-- with that (see Note [FOVArray]) so we just treat it as an opaque blob.
```

**Review by RyanGlScott — APPROVED** [review]

File: saw-core/src/Verifier/SAW/FiniteValue.hs (hunk 2 above)
Comment (sauclovian-g) [reply saw-core/src/Verifier/SAW/FiniteValue.hs]: Fixed

File: saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs (hunk 1 above)
Comment (sauclovian-g) [reply saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs]: Fixed

**Comment by sauclovian-g** [comment]
(force-pushed to squash)

**Pull request merged**